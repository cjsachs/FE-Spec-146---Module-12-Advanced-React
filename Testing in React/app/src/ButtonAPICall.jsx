import React from 'react';
import axios from 'axios';

const ButtonAPICall = () => {
  const fetchTodos = async () => {
    const response = await axios.get(
      'https://jsonplaceholder.typicode.com/users/1/todos'
    );
  };
  return (
    <div>
      <h1>{`We're all GOATED`}</h1>
      <button onClick={fetchTodos}>Fetch Todos</button>
    </div>
  );
};
export default ButtonAPICall;
