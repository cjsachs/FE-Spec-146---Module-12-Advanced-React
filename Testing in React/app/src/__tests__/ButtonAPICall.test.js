import React from 'react';
import axios from 'axios';
import { render, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import ButtonAPICall from '../ButtonAPICall';

// mock axios simulates a real api call
jest.mock('axios');

// describe: groups all tests related to the component
describe('Button APICall Component', () => {
  // individual test
  test('fetch api when button is clicked', async () => {
    // mock data that the test will use to simulate an API response
    const mockResponse = {
      data: [
        {
          userId: 1,
          id: 1,
          title: 'delectus aut autem',
          completed: false,
        },
      ],
    };

    // returns our "mockResponse" when it's called
    axios.get.mockResolvedValue(mockResponse);

    // renders "ButtonAPICall" Componenet
    const { getByText } = render(<ButtonAPICall />);

    // Simulates a "click" event with the text matching "Fetch Todos"
    fireEvent.click(getByText('Fetch Todos'));

    // waits for async operation to complete
    await waitFor(() => {
      // checks that "axios.get" was called on the correct api
      expect(axios.get).toHaveBeenCalledWith(
        'https://jsonplaceholder.typicode.com/users/1/todos'
      );
    });
  });
});

// Snapshot test: Test to ensure that the UI does not change unexpectedly
test('testing snapshot', () => {
  // takes the snapshot
  const { asFragment } = render(<ButtonAPICall/>)
  // expect, toMatchSnapshot: compares the extracted snapshot to the previously stored snapshot
  expect(asFragment()).toMatchSnapshot()
})
