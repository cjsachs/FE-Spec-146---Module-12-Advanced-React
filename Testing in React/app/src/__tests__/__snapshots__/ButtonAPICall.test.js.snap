// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      We're all GOATED
    </h1>
    <button>
      Fetch Todos
    </button>
  </div>
</DocumentFragment>
`;
